accounts/abi/abi.go

```diff
@@ -132,13 +115,61 @@ func (abi ABI) Call(executer Executer, name string, args ...interface{}) interfa
 		return nil
 	}
 
-	output := executer(callData)
+	return abi.parseInterface(name, executer(callData))
+}
 
-	method := abi.Methods[name]
-	ret := make([]interface{}, int(math.Max(float64(len(method.Outputs)), float64(len(output)/32))))
-	for i := 0; i < len(ret); i += 32 {
-		index := i / 32
-		ret[index] = toGoType(method.Outputs[index], output[i:i+32])
+// toGoType parses the input and casts it to the proper type defined by the ABI
+// argument in T.
+func toGoType(i int, t Argument, output []byte) interface{} {
+	index := i * 32
+
+	// Parse the given index output and check whether we need to read
+	// a different offset and length based on the type (i.e. string, bytes)
+	var returnOutput []byte
+	switch t.Type.T {
+	case StringTy, BytesTy: // variable arrays are written at the end of the return bytes
+		// parse offset from which we should start reading
+		offset := int(common.BytesToBig(output[index : index+32]).Uint64())
+		// parse the size up until we should be reading
+		size := int(common.BytesToBig(output[offset : offset+32]).Uint64())
+		// get the bytes for this return value
+		returnOutput = output[offset+32 : offset+32+size]
+	default:
+		returnOutput = output[index : index+32]
+	}
+
+	// cast bytes to abi return type
+	switch t.Type.T {
+	case IntTy:
+		return common.BytesToBig(returnOutput)
+	case UintTy:
+		return common.BytesToBig(returnOutput)
+	case BoolTy:
+		return common.BytesToBig(returnOutput).Uint64() > 0
+	case AddressTy:
+		return common.BytesToAddress(returnOutput)
+	case HashTy:
+		return common.BytesToHash(returnOutput)
+	case BytesTy, FixedBytesTy:
+		return returnOutput
+	case StringTy:
+		return string(returnOutput)
+	}
+	return nil
+}
+
+func (abi ABI) parseInterface(name string, output []byte) interface{} {
+	var (
+		method = abi.Methods[name]
+		ret    = make([]interface{}, len(method.Outputs))
+	)
+
+	if len(ret) > len(output)/32 {
+		return nil
+	}
```

Comment:
Regarding my previous comment, this `if` ensures that the basic parameters are ok, but it cannot check for the dynamic types. Hence we need to return an error or something from the `toGoType` and stop parsing if the case is such.